ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```
@@ -0,0 +1,107 @@
+        * regardless of the ledger effective time of the transaction (e.g. a transaction
+        * with an earlier ledger effective time that gets committed later would find the
+        * contract inactive).
+        */
+      consumedContracts: List[DamlStateKey],
+      /** The contracts created by this transaction.
+        * When the transaction is committed, keys marking the activeness of these
+        * contracts should be created. The key should be a combination of the transaction
+        * id and the relative contract id (that is, the node index).
+        */
+      createdContracts: List[DamlStateKey]
+  )
+
+  def computeInputs(tx: SubmittedTransaction): (List[DamlLogEntryId], List[DamlStateKey]) = {
+    // FIXME(JM): Get referenced packages from the transaction (once they're added to it)
+    def addInput(inputs: List[DamlLogEntryId], coid: ContractId): List[DamlLogEntryId] =
+      coid match {
+        case acoid: AbsoluteContractId =>
+          absoluteContractIdToLogEntryId(acoid)._1 :: inputs
+        case _ =>
+          inputs
+      }
+
+    tx.fold(GenTransaction.TopDown, (List.empty[DamlLogEntryId], List.empty[DamlStateKey])) {
+      case ((logEntryInputs, stateInputs), (nodeId, node)) =>
+        node match {
+          case fetch: NodeFetch[ContractId] =>
+            (addInput(logEntryInputs, fetch.coid), stateInputs)
+          case create: NodeCreate[_, _] =>
+            (logEntryInputs, stateInputs)
+          case exe: NodeExercises[_, ContractId, _] =>
+            (
+              addInput(logEntryInputs, exe.targetCoid),
+              (exe.consuming, exe.targetCoid) match {
+                case (true, acoid: AbsoluteContractId) =>
+                  absoluteContractIdToStateKey(acoid) :: stateInputs
+                case _ =>
+                  stateInputs
+              }
+            )
+          case l: NodeLookupByKey[_, _] =>
+            // FIXME(JM): track fetched keys
```

bail hard! Rather bail than silently succeed. Always log!

---

good point, thanks!